spec/unit/transaction_spec.rb

```diff
@@ -590,6 +590,83 @@ def transaction_with_resource(resource)
 
       transaction.prefetch_if_necessary(resource)
     end
+
+    it "should not prefetch a provider which has failed" do
+      transaction.prefetch_failed_providers[:sshkey][:parsed] = true
+
+      resource.provider.class.expects(:prefetch).never
+
+      transaction.prefetch_if_necessary(resource)
+    end
+
+    describe "and prefetching fails" do
+      before :each do
+        resource.provider.class.expects(:prefetch).raises(Puppet::Error, "message")
+      end
+
+      it "should not rescue prefetch executions, without future_features flag" do
```

Comment:
Did you consider pulling out the future_features as contexts

context "without future_features flag set" do
   before do
     Puppet.settings[:future_features] = false
   end
   ...
end

context "with future_features flag set" do 
  before do
     Puppet.settings[:future_features] = true
   end
   ...
end

Then when this becomes the default we can remove the context and before without having to edit all of the tests

Just a thought.